### DISCUSSION TITLE
👨‍💻☎ WinUI Community Call: April 19, 2023

### DISCUSSION THREAD

**Discussion by gabbybilka** [microsoft/microsoft-ui-xaml#8381] [Announcements]
### https://www.youtube.com/live/Fj4YjgDqm84

# Details
Date: April 19, 2023
Time: 16:00-17:00 UTC (9:00-10:00am Pacific)

Anyone and everyone is welcome - no pre-registration is required.
This will be an informal interactive live stream directly with members of our engineering team.

# Format
The community call is a call among the WinUI team that is live-streamed onto YouTube. We present on new updates, share information, welcome guests, and answer your questions.

### **Agenda**
- Intro
- Photo Gallery in File Explorer
- News & updates in WinAppSDK & WinUI
- Q&A

### **Q&A Code of Conduct**
- To help ensure Q&A remains beneficial for the majority of the call's developer audience, please consider whether your question could be framed so that it might be more broadly relevant to our larger developer audience by ensuring it is reasonably concise, broadly beneficial, well-suited for the fast-paced live Q&A format, or able to spark open ended discussion.  
- Please understand some questions are not well-suited for the fast-paced live Q&A format and should be filed as standalone [Q&A](https://github.com/microsoft/microsoft-ui-xaml/discussions/new?category=q-a&title=%5BYour+Question+Here%5D) discussions where they can be most appropriately handled. Excluding presentation topic requests, this generally includes requests for dedicated follow-ups or technically complex questions that require investigative time or expertise not included in this month's guest list.

### Leave us your questions in the comments on this issue or live in the YouTube chat during the stream!

**Comment by AathifMahir** [comment]
**Reposting from Previous 2 Calls**

Are there any plans for improvement to XAML in General, Xaml has always been the same from the beginning. Regardless of Modern Techniques and etc..

For Example:

- Custom Xaml Binding Like Maui
- Improvements for Compiled Bindings
Etc..

**Comment by mdtauk** [comment]
Any progress on Open Source completion?

When will new controls be coming, and will they bring parity to some missing from WPF, Win32 frameworks to WinUI?

We know Microsoft does not consider a UI designer as essential, but do we have to wait until Build before we learn how you plan to fill in this massive gap.  App Designers are being marginalised from the dev process - whereas Blend was created to allow pure designers a place in the process.

**Comment by ackh** [comment]
Visual Studio's native debug visualization (natvis) for C++/WinRT does not work with custom types in packaged apps out of the box as described [here](https://github.com/microsoft/WindowsAppSDK/issues/2809). Apparently, only the `AppX` folder is searched for `winmd` files but those are not and should not be stored in there. Even if the `winmd` files are copied into the `AppX` folder collections containing custom types do not properly visualize their content in the debugger.

This makes the debugging experience for apps written in C++/WinRT painful. And developing apps without a proper debugging experience should definitely be a thing of the past. Can we expect that this issue is prioritized and addressed soon?

**Reply by pjmlp** [reply]
I bet it will be done after improving the IDL development experience workflow.

**Reply by Scottj1s** [reply]
@ackh, you mention two issues here, both of them known.  Regarding the debug visualizer not knowing about Windows App SDK metadata, I've provided a small MSBuild target as a workaround [here](https://github.com/microsoft/WindowsAppSDK/issues/2809#issuecomment-1512190151).  Regarding collections, there is no [support for visualizing generics](https://github.com/microsoft/cppwinrt/issues/784#issuecomment-724198862).  I looked into it once, and unfortunately it would be a significant investment to implement that.

**Reply by nikolayvpavlov** [reply]
@Scottj1s, I am sorry for revisiting it, but this message "it would be significant investment" is constantly in the air.  While I am perfectly aware that every feature has its costs, the screaming lack of so many features in Windows App Sdk leaves the dev community with the impression that Microsoft considers WASDK a side, small project, with no desire to really invest into it.  The story of Xaml has been very sad: WPF => Silverlight => UWP => WASDK, and none of these got a lot of love after the initial couple of releases. Now, WASDK is so lacking compared to all the rest.  Couple this with the fact that big teams such as Teams and Office (the new Outlook app) choose web stacks than native is also very, very troubling.  

Our company needs to modernize the WPF client of our enterprise LoB app, and the state of WASDK is not giving us comfort.  I really wonder if we should just stay with WPF and employ something like WpfUI instead.

**Reply by pjmlp** [reply]
@nikolayvpavlov You are better off going with the Avalonia folks.

**Reply by nikolayvpavlov** [reply]
@pjmlp, with risk to take it out of topic, I was looking more in the direction of Uno Platform, but maybe it is time I revisit.  x:Bind is awesome and will be missed, though.

**Comment by ghost1372** [comment]
Functions related to UWP and functions that have problems and cause errors were supposed to be removed/hidden from intellisence in version 1.1 ! Why has nothing been done yet?

https://github.com/microsoft/WindowsAppSDK/issues/1660

**Reply by Scottj1s** [reply]
Unfortunately, this issue is complex and there is no quick fix for it.  I added a few comments to the issue.

**Reply by nikolayvpavlov** [reply]
This is also related to issues like [this](https://github.com/microsoft/WindowsAppSDK/issues/1034) which are just not working yet.  Almost two years later there is no improvement.

**Reply by Scottj1s** [reply]
Good discussion and consensus on the related issue - we'll move on this

**Comment by ghost1372** [comment]
The smallest companies and products have a clear and complete roadmap. It's been almost a year that WinUI3/WASDK has no roadmap (publicly). We have the right to know what features the team is currently working on and at what stage it is.

**Reply by BrunoBieri** [reply]
Asked in the November 2022 community call: https://github.com/microsoft/microsoft-ui-xaml/discussions/7881#discussioncomment-4134941

**Reply by ghost1372** [reply]
> Asked in the November community call: https://github.com/microsoft/microsoft-ui-xaml/discussions/7881#discussioncomment-4134941

We ask this question every month but without answer, do you have any answer? I cant remember Nov call.

**Reply by JanRajnoha** [reply]
Maybe best solution here is updating roadmap and not ask every month

**Reply by Panda-Sharp** [reply]
What worries me even more is they had a roadmap on a 3rd party service, than looks like the guy that was maintaining it left and no one else has access to it ... how is this possible in such a big company?

**Reply by nikolayvpavlov** [reply]
@Panda-Sharp this only demonstrates how important the project is for Microsoft.

**Reply by Panda-Sharp** [reply]
@nikolayvpavlov oh "nice", I just noticed this commit
https://github.com/microsoft/microsoft-ui-xaml/commit/e472b693dd976642ee52329474fe2492fb9e5316

**Reply by nikolayvpavlov** [reply]
@Panda-Sharp , and the document that is referenced is about 2021 and 2022.  Awesome.  Just awesome.

**Comment by ghost1372** [comment]
Why does no one care about PublishSingleFile issue?
Almost a year has passed since the report and nothing has happened.

https://github.com/microsoft/microsoft-ui-xaml/issues/7143

**Reply by Scottj1s** [reply]
@ghost1372 you're right - that's a long time to wait for a fix.  I've self-assigned and am investigating.

**Reply by ghost1372** [reply]
Thank you

**Comment by nikolayvpavlov** [comment]
What is the progress of WinUI 3 Xaml Islands?  That would be a great tool for application modernization.

**Comment by JeganRavichandran** [comment]
I cant find any roadmap for the year 2023 or 2024.
What are the high-level or major plans for upcoming release?
If you already have this, please share me a link.

**Reply by nikolayvpavlov** [reply]
The roadmap shared during last few Community Calls is too generic and vague.  I would appreciate a more detailed roadmap, that includes progress on new controls, stability improvements, and performance improvements, esp. in rendering.

**Reply by riverar** [reply]
Opinion: The WinUI team talked about focusing primarily on first-party / partner scenarios (e.g. File Explorer, Apple, Adobe, etc.) I don't think we'll see much community engagement until 2024.

**Comment by mymbrooks** [comment]
Any news for WinUI designer? I can't believe there is no designer in VS 2022

**Comment by ackh** [comment]
The [SwapChainPanel](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.controls.swapchainpanel) in Windows App SDK 1.2 has an issue with cleaning up resources as described in #7924. The cleanup system was supposed to be fixed in Windows App SDK 1.3. However, another form of this bug still appears in Windows App SDK 1.3 release 1.3.230331000. Can someone look into #7924 again please?

**Reply by Scottj1s** [reply]
@ackh The latest stacks are a bit different, but I've reopened the issue for further investigation.

**Comment by mrlacey** [comment]
Is WinAppDriver still connected to this team? If so can we have an update on that? It seems abandoned and with no replacement.

**Reply by marcelwgn** [reply]
Plus one to that, having this as the only option right now seems daunting given its lack of traction and alternative.

**Comment by Diegorro98** [comment]
Last Call I made a question about SVG as App icons in Windows. And the [answer](https://youtu.be/uM65XgC5bzk?t=2399) I got from Ryan Demopoulos was great!

Some days ago I remembered how the Windows 11 start menu is built, which, as far as I know, it has been rebuilt from zero and maybe it uses XAML, and the start menu from the last 2 Windows version (8 & 10) and I've got some thoughts. 

What about using **XAML** files with paths objects for app icons?
Instead of SVG using for icons, which as Ryan mentioned that could be pretty hard to implement, I think that a XAML path (or multiple paths) could be a great option to substitute icons in start menu, taskbar and even in the new File Explorer (for .exe files) in Windows 11!

It could be added to a project just like Windows 8 & 10 png files for tiles are referenced at Package.appxmanifest (so an additional xaml file is referenced there).

Also, an advantage of using XAML (apart from the fact that using vectorized icons is great option) visual states could be implemented (i.e dark mode visual state).

This is the same way other OS support vectorized icons (so not SVG files).

**Reply by pjmlp** [reply]
Yet another WPF feature not supported in WinUI.

https://www.codeproject.com/Tips/1089489/Vector-Icons-in-WPF

**Reply by Diegorro98** [reply]
@pjmlp I think that you are wrong, I could use a Xaml file with a Path inside an application this way:
Using a page called `page.xaml` with an `StackPanel` (for example) with the `x:Name` property set to `MyStackPanel`. 
Then in page.xaml.cs:
```c#
var icon = File.ReadAllText("C:\\Path\\to\\App1\\App1\\Assets\\icon.xaml");
var iconPath = XamlReader.Load(icon) as UIElement;
MyStackPanel.Children.Add(iconPath);
```
And `icon.xaml` contains the following:
```xml
<Path
    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
    Fill="White" 
    Data="M50 10 L90 30 L90 70 L50 90 L10 70 L10 30 Z" />
```
Then an hexagon is shown on the page's stack panel.
Maybe there are cleaner ways to achieve same result, but it's supported...

**Reply by ryandemopoulos** [reply]
We did briefly look at this, @Diegorro98 , although we didn't really do an in-depth exploration/consideration.

SVG is a huge spec, and XAML Paths are great but certainly not as sophisticated as SVG; no doubt some iconography could be achieved with XAML Paths, and I'm guessing some things can't be done -- again, we didn't go super deep in considering this.  The main reason we didn't explore deeply is just pragmatism: there are tens of thousands of icons in Windows, all in rasterized formats, and even if we have 100% SVG capability + amazing XAML Path support, converting all those just wouldn't have been worth the investment.  So we dropped the consideration.

**Reply by nikolayvpavlov** [reply]
What is worrying is that Microsoft becomes too big, and teams become really detached.  Before, things had way more organized approach.  For example, rich edit has been implemented as a reusable control to be used across many apps from many teams.  Now, the Edge team is doing SVG support in a way that it is usable only through WebView2.  

And don't get me started on how the new Outlook and Teams are now web apps...

**Comment by JanRajnoha** [comment]
I hope WASDK 1.3 is finishing and will be released soon, so I have to ask what the status of problem with input is, where currently input is not working for password boxes - https://github.com/microsoft/microsoft-ui-xaml/issues/6291

Other problem, what I'm currently facing, is problem with option PublishTrimmed, which is not working in my case with .NET 7 - https://github.com/microsoft/WindowsAppSDK/issues/3030

**Reply by JanRajnoha** [reply]
And another related question: How could you close issue about touch input when there are reports about wrong behavior even from experimental release. I think no one even looked at it and you just released it and said okay. Could we get some explanation why this happened?

**Comment by horaciodiez** [comment]
Hi, I am a little concerned about the lack of WinUI presence at Build. Only one session about the migration of File Explorer to WinUI.
What can you say about this?

**Comment by lukeblevins** [comment]
I would like to ask Microsoft whether it's reasonable to build apps with WinUI 3 given the **lack of investment** and **no published roadmap**. 

Additionally:
* Is WinUI 3 considered a viable and future-proof technology for Windows app development? 
* What are the plans and goals for WinUI 3 in the near and longer term? 
* How can those in the software industry pitch WinUI 3 given the lack of forward guidance?
* Since there has been continuous under-investment in the UI layer, why is it valuable to open issues on GitHub?
* Given that selling points of WinUI and project "reunion" were never delivered on, can you clarify what changed in Microsoft's stance on native app development in 2020?
* What do measurements of how successful WinUI as a project is look like in 2023?

**Comment by AathifMahir** [comment]
Any plans to make Microsoft.Extension.DependencyInjection to Support WinUI 3 by Default? Maybe Bundling with Visual Studio Starter Template?

**Comment by ghost1372** [comment]
You really embarrass us every day more than yesterday, we wait for several months and in the end we don't get anything and we face some simple and trivial features.
Many basic features and facilities are not yet available!  Recently, version 1.3 was released and it had nothing new!  The only thing there was to change the functionality of the systembackdrop and the appwindow. I regret spending my time learning and developing WinUI 3. We are in 2023 and still the new technology of WinUI 3 uses old and problematic technologies (Old xaml, no designer and more). You are the software giant of the world, hire some more engineers. I wish I spent my time on another technology.No one cares about the issues, no one answers the issues.  And the only answer we hear from you is soon!  When is your "soon" coming? Learn from Telegram, every update is full of new and innovative features!

**Reply by JanRajnoha** [reply]
Develop new technology is not simple process and need huge amount of time. You could hire more engineers, but more people is not guarantee of faster development, this is simple manager fact. Problem with new version 1.3 is with communication. During comunnity calls people from Ms many times mention, thus update is mainly for stability and inner updates to make it better, but it won't have many new functions. Yeah, communication is still biggest problem of MS in windows app sdk. Okay, so let's say, this is maintenance update and MS could not hire more engineers. But what about xaml and issues? Xaml is long term problem which people want to be solved, but again according to MS, they are getting feedback, because it is extremely expensive proces. However, I want it too. Xaml I think is okay, but missing features from wpf ages. And issues... MS promised to make backlog smaller, but nothing happened and I'm angry too, we could just hope more will be assigned to resolve them.

I'm working with winui and windows app sdk more than two years and after that time I could make whole system for my bar. Sometimes it has been pain, but everytime I found solution and currently there is not many problems that need to be solved asap. I think this technology will be  ready for production software soon.

**Reply by ghost1372** [reply]
Thank you very much for your nice explanation, but it has been several years since WASDK was introduced! Many existing technologies (Maui/Uno) were introduced almost at the same time as WASDK and now for resizing a window we should call p/invoke! The development speed is very slow and weak. I hope the situation will change

**Reply by JanRajnoha** [reply]
Yeah, i agree with slow development and other technologies are far than WASDK, but this is problem of closed source and maybe inside problems of MS. I don't want to say MS is great, in last years they prove they are not, but I believe, they do maximum to make WASDK great, because if they fail, I don't know, what other desktop developers will use, but what I know from MS it will be nothing, because everything is old or discontinued. And I hope too, because I don't want to rewrite x apps because they fail :/

**Reply by pjmlp** [reply]
Small history lesson from WinRT development.

In 2012, Windows 8 got released, and we were supposed to migrate .NET and Win32 applications into WinRT, using .NET Native and C++/CX. Interesting COM evolution that pinged back into Longhorn rewrite from .NET back into COM, as internal Microsoft politics rebooted the Longhorn project into Vista.

At first sight, quite interesting, somehow picking up on the Ext-VOS project that gave birth to .NET, however probably as it should have been all along.

However not only were we supposed to rewrite our applications, we had to do it three times, as the WinRT model introduced in Windows 8, required three separate versions of the XAML code for mobile, tablet and desktop.

The year thereafter, Windows 8.1 comes out, the UAP programming model gets introduced, those that already invested into WinRT, had to rewrite their XAML views yet again, as the UAP model merged mobile and tablet modes into a single one, now only two variants were required.

A couple of years later in 2015, Windows 10 gets introduced, we are sold into the One Core vision, UAP gets replaced with UWP, .NET Native changes the compilation model from MDIL (from Singularity/Bartok) into AOT (Midori/Project N), we have yet again to rewrite our XAML code into the new One Core vision for UWP.

In the process, C++/CX gets depreacted, replaced by C++/WinRT with a tooling experience remeniscient from doing ATL in Visual C++, back in yearly 2000, while at CppCon 2016 we get the promise that tooling was in the works to give us back the same development experience of C++/CX, really soon now (sounds familiar?).

Meanwhile XAML Islands get introduced, XAML Direct, Visual Layer, WinUI comes to be to fill where UWP fails short versus WPF.

When we were finally getting into a good position in UWP with WinUI 2.0, and all the stuff above, Project Reunion gets announced in late 2020.

Now we are in 2023, Project Reunion as it was sold to us is no more, replaced by WinAppSDK/WinUI 3.0, .NET Native joined C++/CX deprecation, while like it happened with C++/WinRT, Native AOT isn't able to provide the same AOT experience for WinRT applications.

Then there is the whole issue those left to dry on Holo Lens (whose Windows 11 recent update focus on WebView 2), and XBox, without any path forward from the now deprecated UWP stack.

The few ones that are still left around, get not only asked yet again to rewrite their applications, to put up with increasing number of broken issues, while hoping that one day the glory days of Windows development with WinRT will be back, even if it takes a couple of more years ("just trust us, this time we will make it").

Either the team doesn't understand how they have burned all their bridges with the Windows  developer community at large, specially those that used to be hardcore advocates for the WinRT platform, or they are too blocked to public admit.

For me I am done with WinRT, rather focus on cross platform stuff, and on the Azure side of Microsoft offerings.

Here is the thing, there are plenty of frameworks for doing WOW-effect applications in Windows, with much better tooling than WinUI will ever have in the near future, and on top of that, cross platform.

An ex-WinRT advocate.

**Reply by Balkoth** [reply]
Worst thing is, the one new feature SystemBackdrop is broken: https://github.com/microsoft/WindowsAppSDK/issues/3572

**Comment by riverar** [comment]
What's left of the Windows development community appears to be generally unhappy with WinUI and Windows App SDK. 

Perhaps it is time to rip off the bandage and concede that there's no real community engagement/pulse left. And instead of keeping around the repository baggage around for _issue tracking_, maybe it's time to just archive/close the WinAppSdk and WinUI repositories and return your focus to delighting your internal customers and close partners, working out of your more familiar private repository.

Just an idea to break out of this toxic loop we appear to be stuck in.

**Reply by gus33000** [reply]
I believe personally there is genuine interest in the community in wanting to use the Windows UI library and many people want the Windows App SDK to improve. I personally think a lot of the core problems could be avoided by providing a roadmap and making more of the project open source (I'm sure a lot of people would be more than welcome to contribute and get this into a better shape).

But I also understand your point of view here, not much open source activity happens on this repository currently (also why aren't these discussions on the appsdk repository to begin with?)

I'm not in favor to close these repositories or archive them, but would love to see more community involvement given there is some demand, from myself and probably from a lot more people. But we cannot do much because the project is not too open source and the roadmap is not very clear, nor the target users at the moment. 

If there was no real community interest left I believe we wouldn't be seeing replies on this thread already, clearly there seems to be a general will to use the windows ui library and make it better

**Reply by Panda-Sharp** [reply]
100% agree with @gus33000 

But I agree with @riverar on this part "What's left of the Windows development community appears to be generally unhappy with WinUI and Windows App SDK."

**Reply by Balkoth** [reply]
A real problem is that this is all closed source without chance to look into it...

**Reply by Pinox** [reply]
@riverar Agree !!  At least provide a migration path forward something like BlazorWebViewControl for WinUI3. Then at least we have an option to start the migration to another MS project that has lots of traction.

Closed sourced projects are dead in this day and age so unless you open source the project that I dont think will ever happen you can just as well encourage migration to something like Blazor.  As least there is then an attempt by MS to not further fragment the community but to unify and consolidate and then internally you have more resources to work on "one thing" that is already open sourced and well supported.

**Reply by saint4eva** [reply]
> What's left of the Windows development community appears to be generally unhappy with WinUI and Windows App SDK.
> 
> Perhaps it is time to rip off the bandage and concede that there's no real community engagement/pulse left. And instead of keeping around the repository baggage around for _issue tracking_, maybe it's time to just archive/close the WinAppSdk and WinUI repositories and return your focus to delighting your internal customers and close partners, working out of your more familiar private repository.
> 
> Just an idea to break out of this toxic loop we appear to be stuck in.

Wooooow, shots fired. Absolutely true.

**Reply by nikolayvpavlov** [reply]
Very well said, @gus33000.  The fact that we are so bitter about the current state only proves that the community needs desperately a stable, well-supported and modern platform for native application development.  Despite Microsoft dropping ball with the mobile platform (which spelled the death of UWP), 2-in-1s are still a factor, 2-display laptops are emerging, and Win32 and WPF don't quite fit there.  Further, even enterprises are becoming interested in having their UI / UX modernized even for LoB apps.  But for these apps we need a stable platform that is RAD-based, not something half-baked.

**Reply by Panda-Sharp** [reply]
@nikolayvpavlov I like that you mentioned "2-in-1s are still a factor, 2-display laptops are emerging" and "even enterprises are becoming interested in having their UI / UX modernized even for LoB apps" ... But sadly, we're witnessing MS moving their inbox apps, OS controls and resources to web tech instead of dogfooding their own superior (fast, light, touch&pen friendly) native tech.

I forgot when it was last time I used touch&pen on my 2in1 since Whiteboard, Photos (the editor), Mail (outlook), ecc... moved to web tech.

**Reply by lukeblevins** [reply]
> What's left of the Windows development community appears to be generally unhappy with WinUI and Windows App SDK. 
> 
> 
> 
> Perhaps it is time to rip off the bandage and concede that there's no real community engagement/pulse left. And instead of keeping around the repository baggage around for _issue tracking_, maybe it's time to just archive/close the WinAppSdk and WinUI repositories and return your focus to delighting your internal customers and close partners, working out of your more familiar private repository.
> 
> 
> 
> Just an idea to break out of this toxic loop we appear to be stuck in.

Toxic indeed. Developers have been strung along for far too long and keeping this repo open without delivering on the vision of transparency seems like a mistake. 

As someone who is passionate about bringing the whole of the platform forward for all developers, I'm not sure what the strategy is here.

**Reply by nikolayvpavlov** [reply]
> I forgot when it was last time I used touch&pen on my 2in1 since Whiteboard, Photos (the editor), Mail (outlook), ecc... moved to web tech.

Sigh.  Whiteboard is sluggish as a pregnant snail, the new Outlook is slower than the native version but consumes twice the RAM.  Just splendid, Microsoft.

**Comment by saint4eva** [comment]
Whenever open sourcing the repositories is mentioned, the team will dodge the question.

**Reply by mdtauk** [reply]
I wonder if turnover within the WinUI division is to blame for the shifting promise for Open Sourcing.

**Reply by saint4eva** [reply]
Then the community can help them.

**Reply by koenvd** [reply]
I read the discussion as "wait on our announcement on Build" 🤞 .

**Reply by saint4eva** [reply]
That was what it seemed. I hope they open sourced it so a lot of bugs and fixes will be made by the community.

**Reply by AathifMahir** [reply]
Hope, we'd have the Open Sourced Version of Windows App SDK by the end of this year, it would be great since many devs still eagerly waiting to contribute the project

**Reply by saint4eva** [reply]
And also, all the common features should be moved from WinUI to WASDK so others who would like to build a UI library without needing WInUI can do so.

**Comment by bogdan-patraucean** [comment]
@gabbybilka any update for the may community call? :)

**Reply by gabbybilka** [reply]
Due to Build we're pushing our call out to June, just posted the discussion for Q&A 😊 https://github.com/microsoft/microsoft-ui-xaml/discussions/8487
